### DISCUSSION TITLE
How soon will be realeased new version for .NET 9?

### DISCUSSION THREAD

**Discussion by PavelNavitski1988** [emgucv/emgucv#972] [General]
I updated the solution to .NET 9 and sometimes my application crashes within 20 minutes of working with the IP camera:
Application: Operator.Application.exe
CoreCLR Version: 9.0.24.52809
.NET Version: 9.0.0
Description: The process was terminated due to an unhandled exception.
Stack:
   at [Emgu.CV](http://emgu.cv/).CvInvoke.cveVideoCaptureGrab(IntPtr)
   at [Emgu.CV](http://emgu.cv/).CvInvoke.cveVideoCaptureGrab(IntPtr)
   at [Emgu.CV](http://emgu.cv/).VideoCapture.Run(Emgu.Util.ExceptionHandler)
   at [Emgu.CV](http://emgu.cv/).VideoCapture+<>c__DisplayClass36_0.<Start>b__0()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(System.Threading.Thread, System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef, System.Threading.Thread)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool+WorkerThread.WorkerThreadStart()

and

Faulting application name: Operator.Application.exe, version: 3.0.0.0, time stamp: 0x67200000
Faulting module name: coreclr.dll, version: 9.0.24.52809, time stamp: 0x672049fc
Exception code: 0xc0000005
Fault offset: 0x00000000002eea8b
Faulting process id: 0x714
Faulting application start time: 0x01db3a6193524074
Faulting application path: C:\Program Files\Operator Application\Operator.Application.exe
Faulting module path: C:\Program Files\dotnet\shared\Microsoft.NETCore.App\9.0.0\coreclr.dll
Report Id: 83dd6079-f136-4d34-a535-e0a7dabe5cfb
Faulting package full name:
Faulting package-relative application ID

I think the library is not ready to work with .NET 9

**Comment by PavelNavitski1988** [comment]
I using emgu.cv version 4.9.0.5494
```
public void StartGetIpCameraVideoStream()
{
try
{
if (_videoCapture == null)
{
var cameraUrl = CameraType switch
{
eCameraType.Dahua => $"rtsp://{Login}:{Password}@{IpAddress}:{Port}",
eCameraType.Vivotek => $"rtsp://{Login}:{Password}@{IpAddress}:{Port}/{StreamAccessName}",
_ => throw new Exception("Wrong camera type")
};

        using var ctSource = new CancellationTokenSource(3000);
        var token = ctSource.Token;

        var task = Task.Run(() =>
        {
            _videoCapture = new VideoCapture(cameraUrl, VideoCapture.API.Ffmpeg);

            if (token.IsCancellationRequested)
            {
                Stop();
            }
        }, token);
        task.Wait(token);

        _videoCapture.ImageGrabbed += VideoCaptureImageGrabbed;
    }

    _videoCapture.Start();

    logger.Trace($"Video stream started for camera {IpAddress}");
}
catch (Exception ex)
{
    Stop();

    logger.Trace(ex);
}
}

private void VideoCaptureImageGrabbed(object sender, EventArgs e)
{
try
{
using var mat = new Mat();

    if (!_isSelected || _videoCapture?.Retrieve(mat) != true)
    {
        logger.Trace($"Couldn't get a frame from the camera {IpAddress}");

        VideoFrame = null;
    }
    else
    {
        using var bitmap = mat.ToBitmap();
        VideoFrame = ImageHelper.BitmapToBitmapSource(bitmap);
    }
}
catch { }
}
```

**Comment by emgucv** [comment]
The existing Emgu.CV & Emgu.CV.runtime.windows nuget package is build for .NetStandard 2.0. It is compatible with both .Net 8.0 and .Net 9.0. 

I believe the bug you are experiencing will be consistence either targeting .Net 8.0 / 9.0. 

>> VideoFrame = ImageHelper.BitmapToBitmapSource(bitmap);

This line above, the old BitmapSource in VideoFrame is not released before it is replaced, which may cause memory issues. BitmapSource is IDisposable and you will need to handle it carefully.

**Reply by PavelNavitski1988** [reply]
BitmapSource is not IDisposable interface, what are you talking about??

**Reply by emgucv** [reply]
How do you implement your ImageHelper.BitmapToBitmapSource function?

**Reply by PavelNavitski1988** [reply]
```
public static BitmapSource BitmapToBitmapSource(Bitmap bitmap)
{
    if (bitmap == null)
    {
        return null;
    }

    var bitmapImage = new BitmapImage();

    try
    {
        using var stream = new MemoryStream();
        bitmap.Save(stream, ImageFormat.Bmp);

        bitmapImage.BeginInit();
        bitmapImage.StreamSource = stream;
        bitmapImage.CacheOption = BitmapCacheOption.OnLoad;
        bitmapImage.EndInit();

        if (bitmapImage.CanFreeze)
        {
            bitmapImage.Freeze();
        }
    }
    catch (Exception ex)
    {
        logger.Error(ex);
    }

    return bitmapImage;
}
```

**Reply by emgucv** [reply]
Can you reproduce the problem with a small project and upload it here?

**Reply by PavelNavitski1988** [reply]
The problem is floating and there are no steps to reproduce it. We tested it on a laptop with Wi-Fi. How to start:
1) Build app
2) Start
3) After created file "inputCameraUrl.txt" in .exe folder edit rtsp camera url
4) Start

[WpfApp1.zip](https://github.com/user-attachments/files/17826574/WpfApp1.zip)

**Reply by PavelNavitski1988** [reply]
Perhaps due to bad Wi-Fi, at some point inside the library there is a failure and the entire application crashes, but this should not happen, the application should not die, it should at least swear, or the video stream should stop, but not a complete crash of the application

**Comment by PavelNavitski1988** [comment]
solved the problem by replacing it with another library: https://github.com/nager/Nager.VideoStream

**Reply by jennHarper** [reply]
Thank you for the suggestion.  I am having the same problem.
